zerver/models.py

```diff
@@ -768,11 +762,55 @@ class RealmFilter(models.Model):
     id: int = models.AutoField(auto_created=True, primary_key=True, verbose_name='ID')
     realm: Realm = models.ForeignKey(Realm, on_delete=CASCADE)
     pattern: str = models.TextField(validators=[filter_pattern_validator])
-    url_format_string: str = models.TextField(validators=[URLValidator(), filter_format_validator])
+    url_format_string: str = models.TextField()
+
+    group_match_regex = re.compile(r'%\((?P<group_name>[^)]+)\)s')
 
     class Meta:
         unique_together = ("realm", "pattern")
 
+    def clean(self) -> None:
+        # Django's `full_clean` calls `clean_fields` followed by `clean` and
+        # stores all ValidationErrors from all stages to return as JSON. Thus,
+        # these validations are in addition to field specific validations.
+
+        # extract substitutions
+        try:
+            pattern = re.compile(self.pattern)
+        except Exception:
+            # regex validation errors would be caught by full_clean()
+            # so we can safely just return here.
+            return
+        group_list = list(pattern.groupindex.keys())
+
+        # Report regexes without any patterns.
+        if len(group_list) == 0:
+            raise ValidationError(_("No groups found in URL format string."))
+
+        # Report patterns missing in filter pattern.
+        for m in self.group_match_regex.finditer(self.url_format_string):
+            name = m.group('group_name')
+            if name not in group_list:
+                raise ValidationError(_("Group '{}' in URL format string is "
+                                        "not present in filter pattern.".format(name)))
```

Comment:
Incorrect use of `_()`; the `.format` needs to be outside. I wonder how you evaded our semgrep check for this?